---
title: "Poplar CRISPR lines Differential Expression"
subtitle: "u2023008"
author: "Fai Theerarat Kochakarn"
date: "`r Sys.Date()`"
output:
 html_document:
   fig_width: 9
   fig_height: 6
   toc: true
   number_sections: true
   toc_depth: 3
   toc_float:
     collapsed: TRUE
     smooth_scroll: TRUE
   code_folding: hide
   theme: "flatly"
   highlight: pygments
   includes:
     before_body: header.html
     after_body: footer.html
   css: style.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<hr />
&nbsp;

# Setup
The first two sections are for reproducibility purposes. The detail of the result is shown in section 3 (Quality Control).

* Libraries
```{r}
suppressPackageStartupMessages({
    library(data.table)
    library(DESeq2)
    library(gplots)
    library(here)
    library(hyperSpec)
    library(RColorBrewer)
    library(tidyverse)
    library(VennDiagram)
    library(emoji)
    library(ggVennDiagram)
})
```

* Helper files
```{r}
suppressMessages({
    source(here("UPSCb-common/Rtoolbox/src/plotEnrichedTreemap.R"))
    source(here("UPSCb-common/src/R/featureSelection.R"))
    source(here("UPSCb-common/src/R/volcanoPlot.R"))
		source(here("UPSCb-common/src/R/gopher.R"))
		source(here("UPSCb-common/src/R/topGoUtilities.R"))
})
```

* Graphics
```{r}
pal=brewer.pal(8,"Dark2")
hpal <- colorRampPalette(c("blue","white","red"))(100)
mar <- par("mar")
```

* Functions
1. plot specific gene expression
```{r edit1, echo=FALSE,eval=FALSE}
CHANGEME - here you need to change the variables in the 
plot to display the expression values accross your samples
The example below has 2 variables MGenotype and MDay. These 
need replacing by the variable(s) of interest in your project
```

```{r}
"line_plot" <- function(dds=dds,vst=vst,gene_id=gene_id){
    #message(paste("Plotting",gene_id))
    sel <- grepl(gene_id,rownames(vst))
    stopifnot(sum(sel)==1)

    p <- ggplot(bind_cols(as.data.frame(colData(dds)),
                          data.frame(value=vst[sel,])),
                aes(x=Genotype,y=value,fill=Genotype)) +
        geom_boxplot() +
        geom_jitter(color="black") +
        scale_y_continuous(name="VST expression") + 
        ggtitle(label=paste("Expression for: ",gene_id))
    
    suppressMessages(suppressWarnings(plot(p)))
    return(NULL)
}
```

2. extract the DE results. Default cutoffs are from Schurch _et al._, RNA, 2016
```{r}
"extract_results" <- function(dds,vst,contrast,
                              padj=0.01,lfc=0.5,
                              plot=TRUE,verbose=TRUE,
                              export=TRUE,default_dir=here("data/analysis/DE"),
                              default_prefix="DE-",
                              labels=colnames(dds),
                              sample_sel=1:ncol(dds),
                              expression_cutoff=0,
                              debug=FALSE,filter=c("median",NULL),...){
    
    # get the filter
    if(!is.null(match.arg(filter))){
        filter <- rowMedians(counts(dds,normalized=TRUE))
        message("Using the median normalized counts as default, set filter=NULL to revert to using the mean")
    }
    
    # validation
    if(length(contrast)==1){
        res <- results(dds,name=contrast,filter = filter)
    } else {
        res <- results(dds,contrast=contrast,filter = filter)
    }
    
    stopifnot(length(sample_sel)==ncol(vst))
    
    if(plot){
        par(mar=c(5,5,5,5))
        volcanoPlot(res)
        par(mar=mar)
    }
    
    # a look at independent filtering
    if(plot){
        plot(metadata(res)$filterNumRej,
             type="b", ylab="number of rejections",
             xlab="quantiles of filter")
        lines(metadata(res)$lo.fit, col="red")
        abline(v=metadata(res)$filterTheta)
    }
    
    if(verbose){
        message(sprintf("The independent filtering cutoff is %s, removing %s of the data",
                        round(metadata(res)$filterThreshold,digits=5),
                        names(metadata(res)$filterThreshold)))
        
        max.theta <- metadata(res)$filterNumRej[which.max(metadata(res)$filterNumRej$numRej),"theta"]
        message(sprintf("The independent filtering maximises for %s %% of the data, corresponding to a base mean expression of %s (library-size normalised read)",
                        round(max.theta*100,digits=5),
                        round(quantile(counts(dds,normalized=TRUE),probs=max.theta),digits=5)))
    }
    
    if(plot){
        qtl.exp=quantile(counts(dds,normalized=TRUE),probs=metadata(res)$filterNumRej$theta)
        dat <- data.frame(thetas=metadata(res)$filterNumRej$theta,
                          qtl.exp=qtl.exp,
                          number.degs=sapply(lapply(qtl.exp,function(qe){
                              res$padj <= padj & abs(res$log2FoldChange) >= lfc & 
                                  ! is.na(res$padj) & res$baseMean >= qe
                          }),sum))
        if(debug){
            plot(ggplot(dat,aes(x=thetas,y=qtl.exp)) + 
                     geom_line() + geom_point() +
                     scale_x_continuous("quantiles of expression") + 
                     scale_y_continuous("base mean expression") +
                     geom_hline(yintercept=expression_cutoff,
                                linetype="dotted",col="red"))
        
            p <- ggplot(dat,aes(x=thetas,y=qtl.exp)) + 
                geom_line() + geom_point() +
                scale_x_continuous("quantiles of expression") + 
                scale_y_log10("base mean expression") + 
                geom_hline(yintercept=expression_cutoff,
                           linetype="dotted",col="red")
            suppressMessages(suppressWarnings(plot(p)))
            
            plot(ggplot(dat,aes(x=thetas,y=number.degs)) + 
                     geom_line() + geom_point() +
                     geom_hline(yintercept=dat$number.degs[1],linetype="dashed") +
                     scale_x_continuous("quantiles of expression") + 
                     scale_y_continuous("Number of DE genes"))
            
            plot(ggplot(dat,aes(x=thetas,y=number.degs[1] - number.degs),aes()) + 
                     geom_line() + geom_point() +
                     scale_x_continuous("quantiles of expression") + 
                     scale_y_continuous("Cumulative number of DE genes"))
            
            plot(ggplot(data.frame(x=dat$thetas[-1],
                                   y=diff(dat$number.degs[1] - dat$number.degs)),aes(x,y)) + 
                     geom_line() + geom_point() +
                     scale_x_continuous("quantiles of expression") + 
                     scale_y_continuous("Number of DE genes per interval"))
            
            plot(ggplot(data.frame(x=dat$qtl.exp[-1],
                                   y=diff(dat$number.degs[1] - dat$number.degs)),aes(x,y)) + 
                     geom_line() + geom_point() +
                     scale_x_continuous("base mean of expression") + 
                     scale_y_continuous("Number of DE genes per interval"))
            
            p <- ggplot(data.frame(x=dat$qtl.exp[-1],
                                   y=diff(dat$number.degs[1] - dat$number.degs)),aes(x,y)) + 
                geom_line() + geom_point() +
                scale_x_log10("base mean of expression") + 
                scale_y_continuous("Number of DE genes per interval") + 
                geom_vline(xintercept=expression_cutoff,
                           linetype="dotted",col="red")
            suppressMessages(suppressWarnings(plot(p)))
        }
    }
    
    sel <- res$padj <= padj & abs(res$log2FoldChange) >= lfc & ! is.na(res$padj) & 
        res$baseMean >= expression_cutoff
    
    if(verbose){
      message(sprintf(paste(
        ifelse(sum(sel)==1,
               "There is %s gene that is DE",
               "There are %s genes that are DE"),
        "with the following parameters: FDR <= %s, |log2FC| >= %s, base mean expression > %s"),
        sum(sel),padj,
        lfc,expression_cutoff))
    }
    
    # proceed only if there are DE genes
    if(sum(sel) > 0){
        val <- rowSums(vst[sel,sample_sel,drop=FALSE])==0
        if (sum(val) >0){
          warning(sprintf(paste(
            ifelse(sum(val)==1,
                   "There is %s DE gene that has",
                   "There are %s DE genes that have"),
            "no vst expression in the selected samples"),sum(val)))
          sel[sel][val] <- FALSE
        } 

        if(export){
            if(!dir.exists(default_dir)){
                dir.create(default_dir,showWarnings=FALSE,recursive=TRUE,mode="0771")
            }
            write.csv(res,file=file.path(default_dir,paste0(default_prefix,"results.csv")))
            write.csv(res[sel,],file.path(default_dir,paste0(default_prefix,"genes.csv")))
        }
        if(plot & sum(sel)>1){
            heatmap.2(t(scale(t(vst[sel,sample_sel]))),
                      distfun = pearson.dist,
                      hclustfun = function(X){hclust(X,method="ward.D2")},
                      trace="none",col=hpal,labRow = FALSE,
                      labCol=labels[sample_sel],...
            )
        }
    }
    return(list(all=rownames(res[sel,]),
                up=rownames(res[sel & res$log2FoldChange > 0,]),
                dn=rownames(res[sel & res$log2FoldChange < 0,])))
}
```

3. extract and plot the enrichment results
```{r}
extractEnrichmentResults <- function(enrichment,
                                     diff.exp=c("all","up","dn"),
                                     go.namespace=c("BP","CC","MF"),
                                     genes=NULL,export=FALSE,
                                     plot=TRUE,export_pdf=TRUE,
                                     default_dir=here("analysis/DE"),
                                     default_prefix="DE"){
    # process args
    diff.exp <- match.arg(diff.exp)
    de <- ifelse(diff.exp=="all","none",
                 ifelse(diff.exp=="dn","down",diff.exp))

    # sanity
    if(is.null(unlist(enrichment)) | length(unlist(enrichment)) == 0){
        message("No GO enrichment for",names(enrichment))
    } else {
        # write out
        if(export){
            write_tsv(bind_rows(enrichment),
                      file=here(default_dir,
                                paste0(default_prefix,"-genes_GO-enrichment.tsv")))
            if(!is.null(genes)){
                write_tsv(
                    enrichedTermToGenes(genes=genes,terms=enrichment$id,url=url,mc.cores=16L),
                    file=here(default_dir,
                              paste0(default_prefix,"-enriched-term-to-genes.tsv"))
                )
            }
        }
      if(plot){
        gocatname <- c(BP="Biological Process",
                       CC="Cellular Component",
                       MF="Molecular Function")
        degname <- c(all="all DEGs",
                     up="up-regulated DEGs",
                     dn="down-regulated DEGs")
        lapply(names(enrichment),function(n){
          lapply(names(enrichment[[n]]),function(de){
            lapply(names(enrichment[[n]][[de]]),function(gocat){
              dat <- enrichment[[n]][[de]][[gocat]]
              if(is.null(dat)){
                message("No GO enrichment for ",degname[de]," in category ",gocatname[gocat])
              } else {
                dat$GeneRatio <- dat$Significant/dat$Annotated
                dat$adjustedPvalue <- parse_double(sub("<","",dat$FDR))
                stopifnot(all(! is.na(dat$adjustedPvalue)))
                dat$Count <- as.numeric(dat$Significant)
                dat <- dat[order(dat$GeneRatio),]
                dat$Term <- factor(dat$Term, levels = unique(dat$Term))
                p <- ggplot(dat, aes(x =Term, y = GeneRatio, color = adjustedPvalue, size = Count)) + 
                  geom_point() +
                  scale_color_gradient(low = "red", high = "blue") +
                  theme_bw() + 
                  ylab("GeneRatio") + 
                  xlab("") + 
                  ggtitle(label=paste("GO enrichment:",degname[de]),
                          subtitle=(gocatname[gocat])) +
                  coord_flip()
                
                print(p)
                
                ggsave(file=here(default_dir,
                                 paste(default_prefix,n,de,gocat,"genes_GO-enrichment.pdf",sep="_")),
                       p)
              }
            })
          })
        })
      }
    }
}
```

* Data
```{r load, echo=FALSE,eval=FALSE}
CHANGEME - here you are meant to load an RData object
that contains a DESeqDataSet object. If you ran the 
biological QA template, you need not change anything
```

```{r}
load(here("analysis/salmon/dds_mergeTechRep.rda"))
dds$Genotype <- relevel(dds$Genotype,ref = "T89")
```

Normalisation for visualisation
```{r}
vsd <- varianceStabilizingTransformation(dds,blind=FALSE)
vst <- assay(vsd)
vst <- vst - min(vst)
#dir.create(here("analysis/DE"),showWarnings=FALSE)
#save(vst,file=here("analysis/DE/vst-aware.rda"))
#write_delim(as.data.frame(vst) %>% rownames_to_column("ID"),
#            here("analysis/DE/vst-aware.tsv"))
```

# Gene of interests
```{r goi, echo=FALSE,eval=FALSE}
CHANGEME - Here, you can plot the expression pattern of your gene of
interest. You need to have a list of genes in a text file, one geneID per line
The ID should exist in your vst data.
Note that for the plot to work, you also need to edit the first function (`line_plot`)
at the top of this file
```

## Knocked-out gene
The targeted locus for line 26 is `Potri.006G174000/Potra001877g14982` (Potra2n6c13821)
And for the line 03, it is `Potri.018G096200/Potra001273g10998` (likely to be Potra2n18c32411)

### Potra2n6c13821 (line26)

* Expression level of targeted gene in all lines
```{r}
kogene <- "Potra2n6c13821"
stopifnot(kogene %in% rownames(vst))
dev.null <- line_plot(dds=dds,vst=vst,gene_id = kogene )
```

`r emoji("point_right")` Both knockout lines have lower expression of **Potra2n6c13821** than in parental T89, and line26 has lower expression than in line3. Please note that, even though the expression levels are significantly lower, there are still read counts on the target gene.

* And here is preliminary statistical test to check if the expression level of the gene after knocking-out is significantly different from the parent. I used two-sided Student's t-Test on the vst expression value

```{r}
exp <- bind_cols(as.data.frame(colData(dds)),data.frame(value=vst[kogene,]))
t.test(exp$value[exp$Genotype == "line26"], exp$value[exp$Genotype == "T89"])
t.test(exp$value[exp$Genotype == "line3"], exp$value[exp$Genotype == "T89"])
```

`r emoji("point_right")`P-values for the hypothesis that the mean expression level of **Potra2n6c13821** in knocked-out lines are not equal to the parent are <0.001 for line26 and <0.01 for line3. 

### Potra2n18c32411 (line3)

* Expression level of targeted gene in all lines
```{r}
kogene <- "Potra2n18c32411"
stopifnot(kogene %in% rownames(vst))
dev.null <- line_plot(dds=dds,vst=vst,gene_id = kogene )
```

`r emoji("point_right")` Line3 has lower expression of **Potra2n18c32411** than in parental T89 and line26.

* And here is preliminary statistical test to check if the expression level of the gene after knocking-out is significantly different from the parent. I used two-sided Student's t-Test on the vst expression value

```{r}
exp <- bind_cols(as.data.frame(colData(dds)),data.frame(value=vst[kogene,]))
t.test(exp$value[exp$Genotype == "line26"], exp$value[exp$Genotype == "T89"])
t.test(exp$value[exp$Genotype == "line3"], exp$value[exp$Genotype == "T89"])
```

`r emoji("point_right")`P-values for the hypothesis that the mean expression level of **Potra2n18c32411** in knocked-out lines are not equal to the parent are <0.0005 for line3. For line26, expression level is not significantly different from T89.

## Other SM and LSM
```{r}
suppressMessages(goi <- read_delim(here("doc/goi.txt")))
knitr::kable(goi)
genelist <- unique(goi$V2[!(goi$V2 == "Potra2n6c13821" | goi$V2 == "Potra2n18c32411")])
genelist <- genelist[genelist %in% rownames(vst)]
dev.null <- lapply(genelist,line_plot,dds=dds,vst=vst)
```

# Differential Expression
```{r}
dds <- DESeq(dds)
```

* Dispersion estimation
The dispersion estimation is adequate
```{r}
plotDispEsts(dds)
```

Check the different contrasts
```{r}
resultsNames(dds)
```

## Results
```{r res, echo=FALSE,eval=FALSE}
CHANGEME - here you need to define the contrast you want to 
study - see the example in the next block. 

The `contrast` can be given
by name, as a list (numerator/denominator) or as a vector of weight (e.g. c(0,1));
read the DESeq2 vignette for more info

The `label` argument is typically one (or a combination) of the metadata stored in colData

The function allows for looking at the independent filtering results using `debug=TRUE`

If you are not satisfied with the default from DESeq2, you can set your own cutoff using `expression_cutoff`

You can change the default output file prefix using `default_prefix`

You can select the set of samples to be added to the `heatmap`, using the `sample_sel` argument. It takes a logical vector.
```

### Line 26
```{r}
line26 <- extract_results(dds=dds,vst=vst,contrast="Genotype_line26_vs_T89",
														default_dir = here("analysis/DE"),
														default_prefix = "Line26_", export=FALSE,
														labels=dds$BioID,
														sample_sel = dds$Genotype %in% c("line26","T89"), 
														cexCol=.9, plot = TRUE, verbose = TRUE)
```


### Line 03
```{r}
line03 <- extract_results(dds=dds,vst=vst,contrast="Genotype_line3_vs_T89",
														 default_dir = here("analysis/DE"),
														 default_prefix = "Line03_", export=FALSE,
														 labels=dds$BioID,
														 sample_sel = dds$Genotype %in% c("line3","T89"), 
														 cexCol=.9, plot = TRUE, verbose = TRUE)
```

## Venn Diagram
```{r venn, echo=FALSE,eval=FALSE}
CHANGEME - Here, you typically would have run several contrasts and you want
to assess their overlap plotting VennDiagrams.

In the examples below, we assume that these resutls have been saved in a list
called `res.list`
```

```{r}
res.list <- list("Line 26"=line26,
								 "Line 03"=line03)
```

### All DE genes
```{r}
ggVennDiagram(lapply(res.list,"[[","all"), label_alpha = 0, label = "count") + 
	scale_fill_gradient(low="white",high = "dodgerblue") + 
	scale_x_continuous(expand = expansion(mult = .1))
```

### Up-regulated DE genes (up in knockout)
```{r}
ggVennDiagram(lapply(res.list,"[[","up"), label_alpha = 0, label = "count") + 
	scale_fill_gradient(low="white",high = "dodgerblue") + 
	scale_x_continuous(expand = expansion(mult = .1))
```

### Down-regulated DE genes (down in knockout)
```{r}
ggVennDiagram(lapply(res.list,"[[","dn"), label_alpha = 0, label = "count") + 
	scale_fill_gradient(low="white",high = "dodgerblue") + 
	scale_x_continuous(expand = expansion(mult = .1))
```

## Gene Ontology enrichment
```{r go, echo=FALSE,eval=FALSE}
Once you have obtained a list of candidate genes, you most probably want
to annotate them.

In the following example, we first identify the background; _i.e._ the
population of expressed genes. We select the genes expressed in a least
2 replicate of one condition at a cutoff of `exp`.

Next we run the enrichment, in the example against `athaliana` using 
the gofer3 REST API (interfaced through the gopher.R script loaded at the
beginning of this fil).

Finally we export the go enrichment as a complete table.
We used to export another table consisting
of only the `id` and `padj` columns for using as input for _e.g._
REVIGO; but since flash is EOL and REVIGO not updated, we instead rely on 
the RtoolBox treemap.

In addition we now also export the list of genes that most likely resulted in
the corresponding go enrichment.

Make sure to change the `url` to match your species
```

Here we used topGO to calculate GO enrichment of DEGs vs all genes. Any term having Benjamini-Hochberg adjusted P-value less than 0.01 is shown.

```{r}
background <- rownames(vst)[featureSelect(vst,dds$Genotype,exp=0.00001)]
goannot <- prepAnnot(mapping = "/mnt/reference/Populus-tremula/v2.2/gopher/gene_to_go.tsv")
PvalCutoff = 0.01
PadjustMethod = "BH"
dir.create(here("analysis/GO"),showWarnings=FALSE,recursive=FALSE)
```


### DEGs from Line26
```{r}
res.list <- list("Line 26"=line26)

suppressMessages(enr.list <- lapply(res.list,function(r){
  lapply(r,topGO,background=background,annotation=goannot,alpha=PvalCutoff,p.adjust=PadjustMethod)
}))

suppressWarnings(invisible(extractEnrichmentResults(enr.list,
                                          default_dir = here("analysis/GO"),
                                          default_prefix = "Line26_")))
```

### DEGs from Line03
```{r}
res.list <- list("Line 03"=line03)

suppressMessages(enr.list <- lapply(res.list,function(r){
  lapply(r,topGO,background=background,annotation=goannot,alpha=PvalCutoff,p.adjust=PadjustMethod)
}))

suppressWarnings(invisible(extractEnrichmentResults(enr.list,
                                          default_dir = here("analysis/GO"),
                                          default_prefix = "Line03_")))
```

### Common DEGs between two lines
```{r}
res.list <- list("CommonLine26andLine03"=list("all"=intersect(line26$all,line03$all),
																							"up"=intersect(line26$up,line03$up),
																							"dn"=intersect(line26$dn,line03$dn)))

suppressMessages(enr.list <- lapply(res.list,function(r){
  lapply(r,topGO,background=background,annotation=goannot,alpha=PvalCutoff,p.adjust=PadjustMethod)
}))

suppressWarnings(invisible(extractEnrichmentResults(enr.list,
                                          default_dir = here("analysis/GO"),
                                          default_prefix = "Common_Line26_Line03_")))
```

### DEGs from Line26 that are NOT overlapped with any DEGs from Line03
```{r}
res.list <- list("ExclusivelyLine26"=list("all"=setdiff(line26$all,line03$all),
																							"up"=setdiff(line26$up,line03$all),
																							"dn"=setdiff(line26$dn,line03$all)))

suppressMessages(enr.list <- lapply(res.list,function(r){
  lapply(r,topGO,background=background,annotation=goannot,alpha=PvalCutoff,p.adjust=PadjustMethod)
}))

suppressWarnings(invisible(extractEnrichmentResults(enr.list,
                                          default_dir = here("analysis/GO"),
                                          default_prefix = "Exclusively_Line26_")))
```

### DEGs from Line03 that are NOT overlapped with any DEGs from Line26
```{r}
res.list <- list("ExclusivelyLine03"=list("all"=setdiff(line03$all,line26$all),
																							"up"=setdiff(line03$up,line26$all),
																							"dn"=setdiff(line03$dn,line26$all)))

suppressMessages(enr.list <- lapply(res.list,function(r){
  lapply(r,topGO,background=background,annotation=goannot,alpha=PvalCutoff,p.adjust=PadjustMethod)
}))

suppressWarnings(invisible(extractEnrichmentResults(enr.list,
                                          default_dir = here("analysis/GO"),
                                          default_prefix = "Exclusively_Line03_")))
```

# Session Info 
```{r session info, echo=FALSE}
sessionInfo()
```

